### PR TITLE
[lite] GLM5.2 (DeepSeek-V3.2) IndexShare DSA support

### PR DESCRIPTION
**Pull request by ISEEKYAN**
## GLM5.2 (DeepSeek-V3.2) IndexShare DSA support

Adds GLM5.2 support to Megatron-Lite as a configuration variant of the existing GLM5 (`deepseek_v3_2`) implementation. The model-specific delta is IndexShare plus the GLM5.2 RoPE/context configuration; the DSA implementation remains in the shared attention primitive and reuses the existing fused kernels.

### Implementation

- **IndexShare schedule:** `full` layers compute top-k and following `shared` layers reuse that source. Shared layers have no indexer parameters or indexer loss. The feature is configuration-gated; `index_topk_freq=1` keeps the existing GLM5 all-full path.
- **Packed/CP positions:** the model protocol derives and CP-splits per-sequence positions, and the model/layer/DSA wrapper now forwards those positions unchanged. Packed DSA requires explicit positions; the dense CP fallback uses the shared zigzag position primitive.
- **Pipeline safety:** the split guard covers both trunk and MTP logical layer indices, so a shared MTP layer cannot be placed on a different stage from its full source.
- **Bounded cache lifecycle:** normal forward keeps only the current source group on GPU and drops it after its consumers. Recompute/offload retains old groups on CPU and pages at most one source group back to GPU during backward, then clears the resident group.
- **Checkpoint mapping:** shared layers omit indexer weights on import/export, while full layers retain the existing mapping.

### MTP schedule fact

The official GLM5.2 layout has 78 decoder layers followed by one MTP layer. The MTP transformer is zero-based global layer 78 (one-based layer 79), which is a **full** indexer layer under the `freq=4, offset=3` schedule. It computes its own top-k and does **not** share the last trunk layer's top-k. `index_share_for_mtp_iteration` is retained as source-config compatibility and is not evidence of cross-layer MTP sharing.

### Validation (Slurm GPU, non-skip)

- **Actual causal models vs HF, job 13182986 (`COMPLETED`, `0:0`):** reduced-size four-layer `Glm5Model` fused IndexShare forward, exact MLite-to-HF weight export, and independent `transformers.GlmMoeDsaForCausalLM.forward` over sequence length 1024. Full logits cosine: global `0.9998986721`, token mean `0.9999044538`, token minimum `0.9996886849`, p01 `0.9998022318`; all values are finite. Combined instrumentation for the model forward and following recompute probe recorded `fused_indexer_sparse_attn_with_topk=2`, `dsa_sparse_attn=10`, `indexer_topk=5`.
- **Real protocol packed THD + CP2 + IndexShare + MTP, job 13182986 (`COMPLETED`, `0:0`):** two-GPU `_forward_step` with variable lengths `[16, 20, 24]`, forward and backward, shared trunk layer and full MTP indexer; loss `5.269021e+00`, grad norm `6.683617e+01`, `1 passed` with no skip.
- **Recompute cache lifecycle, job 13182986 (`COMPLETED`, `0:0`):** two source groups through actual checkpoint recomputation; CPU history `4,194,304` bytes, GPU-resident cache entries after forward/backward `0/0`, finite grad norm `0.0194911640`.
- **Fused DSA primitive vs Megatron reference, job 13128510 (`COMPLETED`, `0:0`):** source/shared/MTP attention `max_abs = 0.001953125 / 0.001953125 / 0.0009765625`; Megatron top-k support matched. This is primitive-level parity evidence, not a full-model-logits claim.
- **Weight IO:** save-HF BF16 round trips with IndexShare off/on, job 13128494 (`COMPLETED`, `0:0`), both `max_abs=0.0`.
- **Parallel/runtime regression:** uneven 78-layer pipeline job 13135803, VERL runtime job 13135801, and fused run-to-run acceptance job 13138759 all completed successfully; the non-deterministic fused backward is assessed against deterministic unfused and golden-variance references.
- **Pipeline layout unit matrix:** job 13183070 (`COMPLETED`, `0:0`), `56 passed, 1 xfailed`; the expected xfail is the existing unsupported standalone-MTP placement.

### Focused CPU checks

- Position forwarding, MTP pipeline guard, and cache drop/offload unit checks: `4 passed`.
- Runtime layering and packed bridge checks: `12 passed`.
- `compileall`, targeted Ruff, and `git diff --check` pass.

### CODE REVIEW

**Comment by Meirtz** [comment]
Ran an independent correctness pass on this PR (head 097ea4fca) on B100 (SM100): freq=1 no-regression vs main 69ea18d07, IndexShare lifecycle axes (microbatch loop, recompute/offload paging of the top-k cache), fixed-weight pp1↔pp2 / cp1↔cp2 / ep1↔ep2 invariance on tiny GLM5.2 configs (8L freq=4/offset=3, and 6L+MTP), HF-export checkpoint roundtrip, and an mlite→HF cross-check against `transformers.GlmMoeDsaForCausalLM`.

Everything landed at the fused-kernel run-to-run noise floor (loss rel ~1e-4, grad_norm rel ~1e-3); HF forward logits cosine 0.99982 global / 0.99941 per-token min, consistent with the numbers in the PR body. **No correctness defects found.** Four findings from the process:

**1. The two dense-CP2 reference tests cannot run as written.**
`test_glm5_tiny_model_cp2_matches_full_sequence_reference_forward` and `test_glm5_tiny_model_cp2_matches_hf_reference_logits` call `cp_model(hidden_states=...)`, but `Glm5Model.forward` asserts `input_ids is not None` whenever `self.embed` is built (`glm5/lite/model.py:881`), and the test factory always builds embed. Both fail with `AssertionError` on any 2-rank CUDA env. This is pre-existing on main, not introduced here — but they are the model-level validation for the CP rotary reconstruction fix in 097ea4fca, so that fix currently has no runnable model-level test. (The other three tests in the file pass on 2×B100; a fixed-weight cp1↔cp2 model-level compare also agrees within 0.5%, well inside the file's 3e-2 tolerances.)

**2. HF `GlmMoeDsa` does not train the indexer; mlite does.**
In the cross-check, every indexer param on the HF side received no grad (transformers 5.12/5.13 keeps the indexer out of the autograd path), while mlite's indexer backward flows into the shared `q_a` projection. Result: forward logits match (0.9998 cosine) but `q_a_proj`/`q_a_layernorm`/`q_b_proj` grad cosines drop to 0.86–0.93 across all layers. Not a defect in this PR — just worth knowing for anyone aligning training gradients against the HF reference.

**3. `index_share_for_mtp_iteration` is parsed but never consumed.**
`glm5/config.py` stores it (`_HF_FIELDS` + field default) and tests assert its value, but no production code reads it — the MTP layer's indexer type is decided purely by the global layer-number schedule. If upstream GLM5.2 semantics expect this flag to gate MTP top-k reuse, that behavior is silently absent; if not, the field could be dropped.

**4. The bench harness breaks on GLM5 configs.**
- `--keep-experts`: `replace(model_cfg, num_experts=...)` raises `TypeError` because `Glm5Config.num_experts` is a read-only property over `n_routed_experts` (`examples/bench/bench.py:141-151`).
- `--truncate-layers`: truncates `num_hidden_layers` but not `indexer_types`/`mlp_layer_types`, so any real GLM5.2 config.json (78-entry `indexer_types`) fails the `len(indexer_types) must equal num_hidden_layers` validation.

Happy to send small PRs for 1 and 4 if useful.

**Comment by Meirtz** [comment]
Follow-up on findings 1 and 4 — patches inline so they can be folded into the branch directly; happy to send them as PRs instead if you prefer.

**Patch for finding 1** (the dense-CP2 model-level reference test). Two independent blockers: the model requires `input_ids` whenever the embedding is built, and the comparison sliced the SBH output on the wrong dim. Verified on 2×B100: the file goes 2 failed / 3 passed → 1 failed / 4 passed, i.e. `test_glm5_tiny_model_cp2_matches_full_sequence_reference_forward` now actually runs and passes — restoring model-level coverage for the CP rotary reconstruction fix in 097ea4fca.

```diff
--- a/experimental/lite/tests/unit/model/test_glm5_lite_cp_smoke.py
+++ b/experimental/lite/tests/unit/model/test_glm5_lite_cp_smoke.py
@@ -240,13 +240,19 @@ def test_glm5_tiny_model_cp2_matches_full_sequence_reference_forward():
 
     batch, seq = 1, _fused_dsa_seq_len(world)
     torch.manual_seed(100)
-    full_hidden = torch.randn(batch, seq, cfg.hidden_size, device=device, dtype=torch.bfloat16)
-    local_hidden = zigzag_slice_for_cp(full_hidden, rank, world, seq_dim=1).contiguous()
+    # Feed token ids like the sibling smoke test: Glm5Model.forward requires
+    # input_ids whenever the embedding is built, and the embedding lookup is
+    # pointwise, so it commutes with the zigzag slice.  Both models are built
+    # from the same seed, so their embedding tables are identical.
+    full_ids = torch.randint(0, cfg.vocab_size, (batch, seq), device=device)
+    local_ids = zigzag_slice_for_cp(full_ids, rank, world, seq_dim=1).contiguous()
 
     with torch.no_grad():
-        cp_hidden = cp_model(hidden_states=local_hidden)["hidden_states"]
-        ref_hidden = ref_model(hidden_states=full_hidden)["hidden_states"]
-    expected = zigzag_slice_for_cp(ref_hidden, rank, world, seq_dim=1)
+        cp_hidden = cp_model(input_ids=local_ids)["hidden_states"]
+        ref_hidden = ref_model(input_ids=full_ids)["hidden_states"]
+    # hidden_states come back in SBH layout (see the sibling smoke test's
+    # shape assertion), so the zigzag comparison slices dim 0, not dim 1.
+    expected = zigzag_slice_for_cp(ref_hidden, rank, world, seq_dim=0)
 
     torch.testing.assert_close(cp_hidden, expected, atol=1e-1, rtol=1e-1)
```

`test_glm5_tiny_model_cp2_matches_hf_reference_logits` I left unpatched on purpose: while attempting a fix it turned out to need three independent repairs — the same `input_ids` contract, the DeepseekV3-derived state dict carries no `self_attn.indexer.*` weights (loader raises `SafetensorError` on every full-indexer layer), and after synthesizing those it still lacks the MoE expert mapping (`model.layers.1.mlp.experts.0.gate_proj.weight`). Since transformers now ships `GlmMoeDsaForCausalLM` natively (≥5.12, with the indexer and the freq/offset IndexShare schedule), rebuilding this test against it instead of the DeepseekV3 adapter would remove all three problems at once — but that's a design call I'd rather leave to you.

**Patch for finding 4** (bench harness on GLM5 configs). Verified on CPU against both `Glm5Config` (keep_experts=2 / truncate_layers=4 → `_validate` passes, `indexer_types`/`mlp_layer_types` sliced consistently) and `Qwen3MoEConfig` (unchanged behavior).

```diff
--- a/experimental/lite/examples/bench/bench.py
+++ b/experimental/lite/examples/bench/bench.py
@@ -145,8 +145,13 @@ def _make_mlite_model_config_hook(cfg: BenchCliConfig):
                 raise ValueError("keep_experts requires model config with MoE expert metadata.")
             if keep_experts <= 0 or keep_experts > old_num:
                 raise ValueError(f"keep_experts must be in [1, {old_num}], got {keep_experts}.")
+            # glm5 exposes num_experts as a read-only property over the real
+            # field n_routed_experts; replace() only accepts real fields.
+            field_names = {f.name for f in fields(model_cfg)}
+            expert_field = "num_experts" if "num_experts" in field_names else "n_routed_experts"
             return replace(
-                model_cfg, num_experts=keep_experts, num_experts_per_tok=min(old_topk, keep_experts)
+                model_cfg,
+                **{expert_field: keep_experts, "num_experts_per_tok": min(old_topk, keep_experts)},
             )
 
         hooks.append(keep_experts_hook)
@@ -163,10 +168,14 @@ def _make_mlite_model_config_hook(cfg: BenchCliConfig):
                     f"truncate_layers must be in [1, {old_layers}], got {keep_layers}."
                 )
             updates = {"num_hidden_layers": keep_layers}
-            # layer_types is qwen-style metadata; deepseek_v4 / kimi_k2 configs don't carry it.
-            layer_types = getattr(model_cfg, "layer_types", None)
-            if layer_types is not None:
-                updates["layer_types"] = list(layer_types[:keep_layers])
+            # Per-layer metadata lists must keep length == num_hidden_layers:
+            # layer_types (qwen-style), indexer_types / mlp_layer_types (glm5).
+            # A prefix slice of indexer_types stays consistent with the
+            # freq/offset schedule because the schedule is per-layer.
+            for attr in ("layer_types", "indexer_types", "mlp_layer_types"):
+                values = getattr(model_cfg, attr, None)
+                if values is not None:
+                    updates[attr] = list(values[:keep_layers])
             return replace(model_cfg, **updates)
 
         hooks.append(truncate_layers_hook)
```

**Comment by Meirtz** [comment]
Follow-up: full-depth coverage attempt + two environment findings that anyone running this PR on Hopper will want.

**78-layer pp8 run (the in-repo `test_glm52_indexshare_78_layer_uneven_pp_builds_trains_and_keeps_share_groups`, 1×8 H100).** Everything the test asserts *before* its training step passed at true depth: the 78-layer + MTP GLM5.2 config builds at pp=8 through the grouped auto layout, per-rank layer indices are sorted/unique, `validate_dsa_index_share_pipeline_split` passes on every stage, and every local shared layer finds its full source on the same stage. The one-step training that follows died in the toolchain, not in this PR (below). Training-step numerics for the same IndexShare machinery were already verified at the fused-kernel noise floor on SM100 for pp2 / cp2 / ep2 fixed-weight topologies (previous comments), and your packed THD+CP2 test passes on 2×B100; our B100 partitions reject multi-node allocations, so an 8-rank SM100 training run wasn't obtainable on this cluster.

**Environment finding 1 — cutlass-dsl version + a `pip --target` deployment trap (this is what "DSA breaks on Hopper" actually is).**
`nvidia-cudnn-frontend==1.25.0`'s SM90 DSA kernels reference top-level `cutlass.cute.nvgpu.OperandMajorMode`, which is absent from the image's 4.4.1; use `nvidia-cutlass-dsl==4.5.x` — exactly what its `[cutedsl]` extra pins. (4.6.0 does export that symbol, but is still incompatible with these kernels through a different one, `cute.core.ThrMma`; 4.6.0 also moved its package body from `python_packages/` to `dsl_packages/`.) The NGC pytorch:26.04 image ships 4.4.1, where the symbol is absent, so every SM90 kernel dies with `AttributeError` at build time; the SM100 kernels use the fully-qualified `cute.nvgpu.tcgen05.mma.*` path and are fine on 4.4.1 — hence "works on Blackwell, breaks on Hopper". Extra trap for `pip install --target` deployments: `nvidia-cutlass-dsl` exposes its package body via a `.pth` file, which Python only processes in site-packages, never on `PYTHONPATH` — the image's 4.4.1 silently shadows the staged 4.5.0 unless you add `<target>/nvidia_cutlass_dsl/python_packages` to `PYTHONPATH` explicitly. With both fixed, your `test_glm5_dsa_run_to_run_accept_with_proof` passes on H100 (single GPU, fused forward+backward).

**Environment finding 2 — remaining upstream blocker for multi-rank SM90.**
With the stack fixed as above, 8-rank runs on H100 still die with a native `SIGABRT` (`Fatal Python error: Aborted`) inside `cutlass/base_dsl/compiler.py compile` while JIT-compiling `deepseek_sparse_attention/sparse_attention_backward/_interface_sm90.py::flash_attn_bwd_sm90` — deterministically on the same rank across two attempts, surviving a single-process JIT warm-up and a node-local cache dir, while the single-process smoke on the same node passes. Looks like a cudnn-frontend/cutlass-DSL SM90 JIT bug for the kernel variant that rank's stage needs, worth an upstream report; it is not triggered by anything specific to this PR.

Net for this PR: at real 78-layer depth the pp8 grouping/guard machinery holds; the SM90 training-step gap is a toolchain issue, with SM100 covering the training-side evidence.

**Comment by Meirtz** [comment]
Correction to my previous comment's "Environment finding 2": after a deeper investigation I'm **retracting the cudnn-frontend/cutlass-DSL JIT-bug attribution** for the 8-rank SIGABRT.

What the follow-up showed:
- Pre-compiling every packed-path kernel on the main thread before the schedule (a small pytest-plugin warm-up) does **not** stop the abort — it just migrates later (P2P wait, then test teardown). The JIT was a bystander caught on the crashing rank's stack, not the cause.
- `compute-sanitizer memcheck` over the full 8-rank run reports **zero kernel-level invalid accesses** — the DSA kernels themselves are clean.
- The abort tracks a specific slice of *our* cluster environment (older-driver H100 nodes, world=8, packed path) and survives every toolchain permutation we tried (cutlass-dsl 4.5.0/4.5.2, cuda-python 13.2/13.3, larger PG timeouts, serialized execution). Root cause is somewhere in the driver/NCCL native layer of those nodes and is **not** something this PR or the cutedsl stack does wrong.

Still valid from that comment: the `nvidia-cutlass-dsl==4.5.x` pin (4.4.1 lacks `cute.nvgpu.OperandMajorMode`; 4.6.0 has it — our earlier 4.6 sighting was the image's 4.4.1 shadowing via the `.pth` trap — but 4.6.0 still breaks cudnn-frontend 1.25's SM90 kernels via a different symbol, `cute.core.ThrMma`, verified against a properly-isolated 4.6.0 install) and the `.pth`-on-PYTHONPATH deployment trap. One more trap found on the way, for anyone staging deps with `pip --target`: installing `nvidia-cutlass-dsl[cu13]` *with* dependencies pulls the newest `cuda-python` (13.3.x), which shadows the container's matched copy and fails all driver API calls on nodes whose driver is older (`cudaVersion argument (13030) exceeds the driver version (13020)`) — keep `cuda-python` out of the staged target dir and let the container's own version resolve.

While closing this out we also resolved our last open numeric item: an 8-rank `pp2×cp2×ep2` run diverges from a single-rank run of the same fixed weights by ~0.7% loss (traced to the cp×MTP pair) and ~27% grad_norm (traced to the cp×ep pair). Both components reproduce **identically on main without this PR's commits** (`freq=1`, base-tree control), so they are pre-existing engine behavior, not IndexShare-related — noting it here only so nobody re-discovers it against this PR. With that, our verification has no open findings attributable to this PR beyond the items already reported above.
